src/Opaleye/TH.hs

```diff
@@ -0,0 +1,652 @@
+{-# LANGUAGE TemplateHaskell #-}
+{-# LANGUAGE OverloadedStrings  #-}
+
+module Opaleye.TH (
+      makeOpaleyeModels
+    , makeOpaleyeTables
+    , makeAdaptorAndInstances
+    , Options(..)
+    , TableOptions(..)
+    )
+where 
+
+import Database.PostgreSQL.Simple
+import Database.PostgreSQL.Simple.FromField
+import Language.Haskell.TH
+import Language.Haskell.TH.Syntax hiding (lift)
+import Control.Monad.IO.Class
+import Control.Monad
+import Data.Char
+import Data.Maybe
+import Data.List
+import Data.Profunctor.Product.TH (makeAdaptorAndInstance)
+import Data.Profunctor.Product.Default
+import Opaleye
+import Data.Text (Text)
+import Data.Vector (Vector)
+import Data.Typeable
+import Data.Aeson
+import GHC.Int
+
+import Control.Monad.Trans.Class
+import Control.Monad.Trans.Reader
+import Control.Monad.State.Lazy
+
+import Control.Lens
+
+data Options = Options { tableOptions :: [(String, TableOptions)] }
+
+data TableOptions = TableOptions { modelName :: String, overrideDefaultTypes :: [(String, String)], protectedFields :: [String] }
+
+type ColumnType = String
+
+data ColumnInfo = ColumnInfo { columnTableName :: String, columnName ::String, columnType :: ColumnType, columnDefault :: Bool, columnNullable :: Bool, columnPrimary :: Bool} deriving (Show)
+
+type LensClassGenerated = [String]
+type Env = (ConnectInfo, Options, LensClassGenerated)
+type EnvM a = StateT Env Q a
+
+getColumns :: Connection -> String -> IO [ColumnInfo]
```

Comment:
As far as I looked, no, there is no built in pg-simple function to do this. And what sort of edge cases do you have in mind?